File: javascript/src/createMeta.ts
import * as messages from '@cucumber/messages'
import os from 'os'
import { format as formatUrl, parse as parseUrl } from 'url'

import defaultCiDict from './ciDict.json'
import evaluateVariableExpression from './evaluateVariableExpression'
import { CiDict, CiSystem, Env } from './types'

type GitInfo = {
  remote: string
  revision: string
  branch: string
  tag?: string
}

export default function createMeta(
  toolName: string,
  toolVersion: string,
  envDict: Env,
  ciDict?: CiDict
): messages.Meta {
  if (ciDict === undefined) {
    ciDict = defaultCiDict
  }
  return {
    protocolVersion: messages.version,
    implementation: {
      name: toolName,
      version: toolVersion,
    },
    cpu: {
      name: os.arch(),
    },
    os: {
      name: os.platform(),
      version: os.release(),
    },
    runtime: {
      name: 'node.js',
      version: process.versions.node,
    },
    ci: detectCI(ciDict, envDict),
  }
}

export function detectCI(ciDict: CiDict, envDict: Env): messages.Ci | undefined {
  const detected: messages.Ci[] = []
  for (const [ciName, ciSystem] of Object.entries(ciDict)) {
    const ci = createCi(ciName, ciSystem, envDict)
    if (ci) {
      detected.push(ci)
    }
  }
  if (detected.length !== 1) {
    return undefined
  }
  if (detected.length > 1) {
    console.error(
      `@cucumber/create-meta WARNING: Detected more than one CI: ${JSON.stringify(
        detected,
        null,
        2
      )}`
    )
    console.error('Using the first one.')
  }
  return detected[0]
}

export function removeUserInfoFromUrl(value: string): string {
  if (!value) return value
  const url = parseUrl(value)
  if (url.auth === null) return value
  url.auth = null
  return formatUrl(url)
}

function createCi(ciName: string, ciSystem: CiSystem, envDict: Env): messages.Ci | undefined {
  const url = evaluateVariableExpression(ciSystem.url, envDict)
  const buildNumber = evaluateVariableExpression(ciSystem.buildNumber, envDict)
  if (url === undefined) {
    // The url is what consumers will use as the primary key for a build
    // If this cannot be determined, we return nothing.
    return undefined
  }

  const branch = evaluateVariableExpression(ciSystem.git.branch, envDict)
  const tag = evaluateVariableExpression(ciSystem.git.tag, envDict)
  const git: GitInfo = {
    remote: removeUserInfoFromUrl(evaluateVariableExpression(ciSystem.git.remote, envDict)),
    revision: evaluateVariableExpression(ciSystem.git.revision, envDict),
    branch: branch,
  }

  if (tag) {
    git['tag'] = tag
  }

  return {
    name: ciName,
    url,
    buildNumber,
    git: git,
  }
}

File: java/src/main/java/io/cucumber/createmeta/CreateMeta.java
package io.cucumber.createmeta;

import io.cucumber.messages.ProtocolVersion;
import io.cucumber.messages.types.Ci;
import io.cucumber.messages.types.Git;
import io.cucumber.messages.types.Meta;
import io.cucumber.messages.types.Product;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static io.cucumber.createmeta.VariableExpression.evaluate;

public final class CreateMeta {
    private CreateMeta(){

    }

    public static Meta createMeta(
            String implementationName,
            String implementationVersion,
            Map<String, String> env
    ) {
        return new Meta(
                ProtocolVersion.getVersion(),
                new Product(implementationName, implementationVersion),
                new Product(System.getProperty("java.vm.name"), System.getProperty("java.vm.version")),
                new Product(System.getProperty("os.name"), null),
                new Product(System.getProperty("os.arch"), null),
                detectCI(env)
        );
    }

    public static String removeUserInfoFromUrl(String value) {
        if (value == null) return null;
        try {
            URI uri = URI.create(value);
            return new URI(uri.getScheme(), null, uri.getHost(), uri.getPort(), uri.getPath(), uri.getQuery(), uri.getFragment()).toASCIIString();
        } catch (URISyntaxException | IllegalArgumentException e) {
            return value;
        }
    }

    static Ci detectCI(Map<String, String> env) {
        List<Ci> detected = new ArrayList<>();
        for (CiTemplate ciTemplate : CiServers.TEMPLATES) {
            Ci ci = createCi(ciTemplate, env);
            if (ci != null) {
                detected.add(ci);
            }
        }
        return detected.size() == 1 ? detected.get(0) : null;
    }

    private static Ci createCi(CiTemplate ciTemplate, Map<String, String> env) {
        String url = evaluate(ciTemplate.url, env);
        if (url == null) return null;
        String buildNumber = evaluate(ciTemplate.buildNumber, env);
        String remote = removeUserInfoFromUrl(evaluate(ciTemplate.git.remote, env));
        String revision = evaluate(ciTemplate.git.revision, env);
        String branch = evaluate(ciTemplate.git.branch, env);
        String tag = evaluate(ciTemplate.git.tag, env);

        return new Ci(
                ciTemplate.name,
                url,
                buildNumber,
                new Git(remote, revision, branch, tag)
        );
    }
}
